PolyEngine/Core/Src/Optional.hpp

```diff
@@ -0,0 +1,188 @@
+#pragma once
+
+#include "Defines.hpp"
+
+namespace Poly {
+	template<typename V>
+	class Optional {
+	public:
+		using ValueType = V;
+	public:
+		Optional() : initialized(false), dummy{} {}
+		Optional(const V&  value) : initialized(true), value_storage(value) {}
+		Optional(      V&& value) : initialized(true), value_storage(std::move(value)) {}
+		~Optional() { if (this->HasValue()) { this->Value().~V(); } }
```

Comment:
Eh, both your examples do actually provide value on their own :P The argument about debugging/profiling _almost_ convinces me, though I'm left wondering if we have any resources to write such tools that would be better than (or at least as good as) what is already available.